Make API_BASE_PATH, UI_BASE_PATH, UI_EXTERNAL_LOGIN_URI configurable at load-time

himdel opened the pull request:
Closes #93 

The UI builds produced by `npm run build` can be further configured by serving a `/pulp-ui-config.json` alongside the built UI. (Note it has to be mapped at `/`, not just wherever `index.html` is served from.)

* `API_BASE_PATH` - defaults to `/pulp/api/v3/` - change when using domains or a different path
* `UI_BASE_PATH` - defaults to `/ui/` - change when only serving index in a subdirectory, or want different browser path prefix
* `UI_EXTERNAL_LOGIN_URI` - defaults to nothing - set to something like `/login/` when using an SSO
* `EXTRA_VERSION` - an extra version string to display in about modal

Also documents `API_PROXY` in dev mode, and restore SSO logic when `UI_EXTERNAL_LOGIN_URI` is set.

---

default config in `pulp-ui/pulp-ui-config.json` gets copied by `CopyWebpackPlugin` to `dist/`,
and must be served from `/pulp-ui-config.json` in the webserver config for this to work.

`src/ui-config.ts` (imported from `src/entrypoint.tsx`) then loads it and populates a `config` export on success.

entrypoint then ensures we wait for the promise before rendering anything that would need to use that config

Cc @gerrod3

himdel committed:
Make API_BASE_PATH, UI_BASE_PATH, UI_EXTERNAL_LOGIN_URI & EXTRA_VERSION configurable via /pulp-ui-config.json

gerrod3 commented on src/entrypoint.tsx:
```diff
@@ -1,50 +1,95 @@
+        isInline
+        title={t`Error loading /pulp-ui-config.json`}
+      >
+        {error}
```
```suggestion
        {error.toString()}
```

gerrod3 commented on src/ui-config.ts:
```diff
@@ -0,0 +1,5 @@
+export const configPromise = fetch('/pulp-ui-config.json').then((data) =>
+  data.json(),
+);
```
Should we add an error handle for 404? Else the error message on 404 will be JSON parse error.

gerrod3 reviewed (changes requested):
This looks good. I tested it out and we will need to add a simple change to the oci-env, but other than that it works perfectly.

himdel committed:
ui config error - stringify error, handle http errors, add fallback to defaults button

the button will retry with default options (loaded from the same file but at UI build time)

himdel commented:
Thanks, added stringification, http error handling (I forgot `fetch` doesn't reject the promise on 404), and a fallback button to try the default config :)

![20241021235636](https://github.com/user-attachments/assets/90832cfa-67b0-4470-ad7e-beff727eb9fa)

himdel commented:
:crossed_fingers: 

BTW these changes also allow us to change the settings at runtime, so a runtime config UI for these is doable :) Or, say, a domain switcher that just changes UI_BASE_PATH :)
